src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -0,0 +1,429 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @param <R> record type
+ * @param <B> bulk request type
+ */
+public class BulkProcessor<R, B> {
+
+  private static final Logger log = LoggerFactory.getLogger(BulkProcessor.class);
+
+  private final Time time;
+  private final BulkClient<R, B> bulkClient;
+  private final int maxBufferedRecords;
+  private final int batchSize;
+  private final long lingerMs;
+  private final int maxRetries;
+  private final long retryBackoffMs;
+
+  private final Thread farmer;
+  private final ExecutorService executor;
+
+  // thread-safe stats
+  private final AtomicLong createdBatches = new AtomicLong();
+  private final AtomicLong successfulRecords = new AtomicLong();
+  private final AtomicLong successfulBatches = new AtomicLong();
+
+  // thread-safe state, can be mutated safely without synchronization,
+  // but may be part of synchronized(this) wait() conditions so need to notifyAll() on changes
+  private volatile boolean stopRequested = false;
+  private volatile boolean flushRequested = false;
+  private final AtomicReference<ConnectException> error = new AtomicReference<>();
+
+  // shared state, synchronized on (this), may be part of wait() conditions so need notifyAll() on changes
+  private final Deque<R> unsentRecords;
+  private int inFlightRecords = 0;
+
+  public BulkProcessor(
+      Time time,
+      BulkClient<R, B> bulkClient,
+      int maxBufferedRecords,
+      int maxInFlightRequests,
+      int batchSize,
+      long lingerMs,
+      int maxRetries,
+      long retryBackoffMs
+  ) {
+    this.time = time;
+    this.bulkClient = bulkClient;
+    this.maxBufferedRecords = maxBufferedRecords;
+    this.batchSize = batchSize;
+    this.lingerMs = lingerMs;
+    this.maxRetries = maxRetries;
+    this.retryBackoffMs = retryBackoffMs;
+
+    unsentRecords = new ArrayDeque<>(maxBufferedRecords);
+
+    final ThreadFactory threadFactory = makeThreadFactory();
+    farmer = threadFactory.newThread(farmerTask());
+    executor = Executors.newFixedThreadPool(maxInFlightRequests, threadFactory);
+  }
+
+  private ThreadFactory makeThreadFactory() {
+    final AtomicInteger threadCounter = new AtomicInteger();
+    final Thread.UncaughtExceptionHandler uncaughtExceptionHandler = new Thread.UncaughtExceptionHandler() {
+      @Override
+      public void uncaughtException(Thread t, Throwable e) {
+        log.error("Uncaught exception in BulkProcessor thread {}", t, e);
+        failAndStop(e);
+      }
+    };
+    return new ThreadFactory() {
+      @Override
+      public Thread newThread(Runnable r) {
+        final Thread t = new Thread(r, "BulkProcessor-t" + threadCounter.getAndIncrement());
```

Comment:
Do you want to attach more info to the thread name since this won't identify which BulkProcessor they belong to. For example, if you have multiple tasks running in the same process.